File: packages/browser/ios/RNPingBrowserCommon.swift
//
//  RNPingBrowserCommon.swift
//  RNPingBrowser
//
//  Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
//
//  This software may be modified and distributed under the terms
//  of the MIT license. See the LICENSE file for details.
//

import Foundation
import PingBrowser
import PingLogger
import AuthenticationServices
import RNPingCore
import RNPingLogger

private let createQueueKey = DispatchSpecificKey<Void>()

/// Common iOS implementation for the Ping Browser React Native module.
@objcMembers
public class RNPingBrowserCommon: NSObject {
  /// Stable error codes emitted by the Browser module.
  ///
  /// Keep these in sync with JS `BrowserErrorCode` and Android `BrowserErrorCodes`.
  private enum BrowserErrorCode: String {
    case openError = "BROWSER_OPEN_ERROR"
  }

  /// Adapter used to launch and reset the native browser session.
  @MainActor
  internal static var browserLauncher: BrowserLaunching = DefaultBrowserLauncherAdapter()

#if DEBUG
  /// Replace the browser launcher for unit tests.
  @MainActor
  public static func _setBrowserLauncherForTesting(_ launcher: BrowserLaunching) {
    browserLauncher = launcher
  }

  /// Reset the browser launcher after tests.
  @MainActor
  public static func _resetBrowserLauncherForTesting() {
    browserLauncher = DefaultBrowserLauncherAdapter()
  }
#endif

  /// Resolve a native logger from the shared Core logger registry.
  ///
  /// - Parameter loggerId: Logger handle identifier from JS.
  /// - Returns: Native logger instance, or nil when missing/invalid.
  private static func resolveLoggerFromCoreSync(_ loggerId: String?) -> Logger? {
    guard let loggerId, !loggerId.isEmpty else {
      return nil
    }

    guard let handle = RegistrySync.resolveSync(
      loggerId,
      registry: CoreRuntime.loggerRegistry,
      queueKey: createQueueKey,
      context: "RNPingBrowserCommon.resolveLoggerFromCoreSync"
    ) as? LoggerHandleContract else {
      return nil
    }
    return handle.nativeLogger as? Logger
  }

  /// Accepts configuration from JavaScript (currently a no-op on iOS).
  ///
  /// - Parameter config: Optional configuration payload from the JS layer.
  @objc
  public static func configure(_ config: NSDictionary) {
    // Reserved for future iOS customization.
  }

  /// Resets any in-flight browser session.
  ///
  /// This cancels the current browser flow when supported by the native launcher.
  @objc
  public static func reset() {
    Task { @MainActor in
      browserLauncher.reset()
    }
  }

  /// Launches a browser session and reports success, cancel, or error.
  ///
  /// - Parameters:
  ///   - url: The URL to launch in the system browser.
  ///   - options: Launch options including callback scheme, iOS settings, and optional loggerId.
  ///   - resolver: Callback invoked with the result payload.
  ///   - rejecter: Callback invoked for validation or launch errors.
  @objc
  public static func open(
    _ url: String,
    options: NSDictionary,
    resolver: @escaping (NSDictionary) -> Void,
    rejecter: @escaping (String, String, NSError?) -> Void
  ) {
    let handlers = PromiseBridge<NSDictionary>(resolver: resolver, rejecter: rejecter)
    let loggerId = options["loggerId"] as? String

    // Validate required options before launching.
    guard let callbackScheme = options["callbackUrlScheme"] as? String,
          !callbackScheme.isEmpty else {
      let error = GenericError(
        type: .argumentError,
        error: BrowserErrorCode.openError.rawValue,
        message: "callbackUrlScheme is required"
      )
      handlers.reject(error)
      return
    }

    guard let launchUrl = URL(string: url) else {
      let error = GenericError(
        type: .argumentError,
        error: BrowserErrorCode.openError.rawValue,
        message: "Invalid URL"
      )
      handlers.reject(error)
      return
    }
    guard isSupportedHttpUrl(launchUrl) else {
      let error = GenericError(
        type: .argumentError,
        error: BrowserErrorCode.openError.rawValue,
        message: "Unsupported URL scheme. Only HTTP and HTTPS URLs are supported."
      )
      handlers.reject(error)
      return
    }

    let iosOptions = options["ios"] as? NSDictionary
    let browserTypeRaw = iosOptions?["browserType"] as? String
    let browserModeRaw = iosOptions?["browserMode"] as? String

    let browserType: BrowserType
    switch browserTypeRaw {
    case "ephemeralAuthSession":
      browserType = .ephemeralAuthSession
    case "nativeBrowserApp":
      browserType = .nativeBrowserApp
    case "sfViewController":
      browserType = .sfViewController
    default:
      browserType = .authSession
    }
    let browserMode: BrowserMode
    switch browserModeRaw {
    case "logout":
      browserMode = .logout
    case "custom":
      browserMode = .custom
    default:
      browserMode = .login
    }

    Task { @MainActor in
      let nativeLogger = resolveLoggerFromCoreSync(loggerId) ?? LogManager.none

      do {
        let result = try await browserLauncher.launch(
          url: launchUrl,
          customParams: nil,
          browserType: browserType,
          browserMode: browserMode,
          callbackURLScheme: callbackScheme,
          logger: nativeLogger
        )

        handlers.resolve([
          "type": "success",
          "url": result.absoluteString
        ] as NSDictionary)
      } catch let error as BrowserError {
        if case .externalUserAgentCancelled = error {
          handlers.resolve(["type": "cancel"] as NSDictionary)
        } else {
          // Browser errors are mapped to the shared native error contract.
          let mapped = GenericError(
            type: .internalError,
            error: BrowserErrorCode.openError.rawValue,
            message: error.localizedDescription
          )
          handlers.reject(mapped, underlying: error as NSError)
        }
      } catch let error as ASWebAuthenticationSessionError {
        if error.code == .canceledLogin {
          handlers.resolve(["type": "cancel"] as NSDictionary)
        } else {
          let mapped = GenericError(
            type: .internalError,
            error: BrowserErrorCode.openError.rawValue,
            message: error.localizedDescription,
            code: error.code.rawValue
          )
          handlers.reject(mapped, underlying: error as NSError)
        }
      } catch {
        let mapped = GenericError(
          type: .internalError,
          error: BrowserErrorCode.openError.rawValue,
          message: error.localizedDescription
        )
        handlers.reject(mapped, underlying: error as NSError)
      }
    }
  }

  /// Validates that a URL uses an HTTP(S) scheme.
  ///
  /// - Parameter url: URL provided from JavaScript.
  /// - Returns: `true` when the URL scheme is `http` or `https`.
  private static func isSupportedHttpUrl(_ url: URL) -> Bool {
    guard let scheme = url.scheme?.lowercased() else {
      return false
    }
    return scheme == "http" || scheme == "https"
  }
}

File: packages/browser/android/src/main/java/com/pingidentity/rnbrowser/RNPingBrowserCommon.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.rnbrowser

import android.graphics.Color
import androidx.browser.auth.AuthTabColorSchemeParams
import androidx.browser.auth.AuthTabIntent
import androidx.browser.customtabs.CustomTabColorSchemeParams
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.net.toUri
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.WritableMap
import com.pingidentity.browser.BrowserLauncher
import com.pingidentity.browser.BrowserCanceledException
import com.pingidentity.logger.Logger
import com.pingidentity.logger.NONE
import com.pingidentity.rncore.CoreRuntime
import com.pingidentity.rncore.error.ErrorType
import com.pingidentity.rncore.error.GenericError
import com.pingidentity.rncore.error.reject
import com.pingidentity.rncore.logger.LoggerHandleContract
import java.net.MalformedURLException
import java.net.URI
import java.net.URL
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import com.pingidentity.rncore.utils.launchBridge

/**
 * Common utilities for the Ping Browser module.
 */
object RNPingBrowserCommon {

  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
  private var appContext: ReactApplicationContext? = null
  internal var browserLauncher: BrowserLauncherAdapter = DefaultBrowserLauncherAdapter
  internal var mapFactory: () -> WritableMap = { Arguments.createMap() }

  /**
   * Apply global browser configuration from JS.
   */
  @JvmStatic
  fun configure(config: ReadableMap, reactContext: ReactApplicationContext) {
    appContext = reactContext
    val customTabsConfig = if (config.hasKey("customTabs")) {
      config.getMap("customTabs")
    } else {
      null
    }

    val authTabsConfig = if (config.hasKey("authTabs")) {
      config.getMap("authTabs")
    } else {
      null
    }

    val browserPackage = if (config.hasKey("browserPackage")) {
      config.getString("browserPackage")
    } else {
      null
    }

    browserLauncher.customTabsCustomizer = {
      val colorParamsBuilder = CustomTabColorSchemeParams.Builder()
      var hasColors = false
      customTabsConfig?.let { map ->
        if (map.hasKey("showTitle")) {
          setShowTitle(map.getBoolean("showTitle"))
        }
        if (map.hasKey("urlBarHidingEnabled")) {
          setUrlBarHidingEnabled(map.getBoolean("urlBarHidingEnabled"))
        }
        if (map.hasKey("toolbarColor")) {
          val colorValue = map.getString("toolbarColor")
          if (!colorValue.isNullOrBlank()) {
            colorParamsBuilder.setToolbarColor(Color.parseColor(colorValue))
            hasColors = true
          }
        }
        if (map.hasKey("colorScheme")) {
          when (map.getString("colorScheme")?.lowercase()) {
            "light" -> setColorScheme(CustomTabsIntent.COLOR_SCHEME_LIGHT)
            "dark" -> setColorScheme(CustomTabsIntent.COLOR_SCHEME_DARK)
            "system" -> setColorScheme(CustomTabsIntent.COLOR_SCHEME_SYSTEM)
          }
        }
      }
      if (hasColors) {
        setDefaultColorSchemeParams(colorParamsBuilder.build())
      }
    }

    browserLauncher.authTabCustomizer = {
      authTabsConfig?.let { map ->
        if (map.hasKey("ephemeral")) {
          val enabled = map.getBoolean("ephemeral")
          invokeAuthTabBuilderMethod(this, "setEphemeralBrowsingEnabled", enabled)
        }
        if (map.hasKey("colorScheme")) {
          val scheme = when (map.getString("colorScheme")?.lowercase()) {
            "light" -> CustomTabsIntent.COLOR_SCHEME_LIGHT
            "dark" -> CustomTabsIntent.COLOR_SCHEME_DARK
            "system" -> CustomTabsIntent.COLOR_SCHEME_SYSTEM
            else -> null
          }
          if (scheme != null) {
            invokeAuthTabBuilderMethod(this, "setColorScheme", scheme)
          }
        }

        val colorParamsBuilder = AuthTabColorSchemeParams.Builder()
        var hasColors = false
        map.getString("toolbarColor")?.takeIf { it.isNotBlank() }?.let { value ->
          colorParamsBuilder.setToolbarColor(Color.parseColor(value))
          hasColors = true
        }
        map.getString("navigationBarColor")?.takeIf { it.isNotBlank() }?.let { value ->
          colorParamsBuilder.setNavigationBarColor(Color.parseColor(value))
          hasColors = true
        }
        map.getString("navigationBarDividerColor")?.takeIf { it.isNotBlank() }?.let { value ->
          colorParamsBuilder.setNavigationBarDividerColor(Color.parseColor(value))
          hasColors = true
        }
        if (hasColors) {
          invokeAuthTabBuilderMethod(this, "setDefaultColorSchemeParams", colorParamsBuilder.build())
        }
      }
    }

    val resolvedPackage = if (browserPackage.isNullOrBlank()) {
      null
    } else if (isPackageInstalled(browserPackage)) {
      browserPackage
    } else {
      null
    }

    browserLauncher.intentCustomizer = {
      if (!resolvedPackage.isNullOrBlank()) {
        setPackage(resolvedPackage)
      }
    }
  }

  /**
   * Launch a browser session and resolve to success/cancel, or reject on error.
  */
  @JvmStatic
  fun open(url: String, options: ReadableMap, promise: Promise) {
    val loggerId = if (options.hasKey("loggerId")) {
      options.getString("loggerId")
    } else {
      null
    }
    val resolvedLogger = resolveLoggerFromCore(loggerId)
    BrowserLauncher.logger = resolvedLogger ?: Logger.NONE

    val callbackUrlScheme = if (options.hasKey("callbackUrlScheme")) {
      options.getString("callbackUrlScheme")
    } else {
      null
    }
    if (callbackUrlScheme.isNullOrBlank()) {
      promise.reject(
        GenericError(
          type = ErrorType.ARGUMENT_ERROR,
          error = BrowserErrorCodes.BROWSER_OPEN_ERROR,
          message = "callbackUrlScheme is required"
        )
      )
      return
    }

    val redirectUri = if (options.hasKey("redirectUri")) {
      options.getString("redirectUri")
    } else {
      null
    }

    scope.launchBridge(promise, BrowserErrorCodes.BROWSER_OPEN_ERROR) {
      val launchUrl = try {
        parseLaunchUrl(url)
      } catch (e: MalformedURLException) {
        promise.reject(
          GenericError(
            type = ErrorType.ARGUMENT_ERROR,
            error = BrowserErrorCodes.BROWSER_OPEN_ERROR,
            message = e.message
          ),
          e
        )
        return@launchBridge
      }

      val result = try {
        val resolvedRedirectUri = redirectUri?.toUri() ?: browserLauncher.redirectUri
        browserLauncher.launch(launchUrl, resolvedRedirectUri)
      } catch (e: CancellationException) {
        throw e
      } catch (e: Exception) {
        Result.failure(e)
      }

      if (result.isSuccess) {
        val uri = result.getOrNull()
        if (uri == null) {
          val payload = mapFactory()
          payload.putString("type", "cancel")
          promise.resolve(payload)
          return@launchBridge
        }

        val payload = mapFactory()
        payload.putString("type", "success")
        payload.putString("url", uri.toString())
        promise.resolve(payload)
        return@launchBridge
      }

      val error = result.exceptionOrNull()
      if (error is BrowserCanceledException) {
        val payload = mapFactory()
        payload.putString("type", "cancel")
        promise.resolve(payload)
        return@launchBridge
      }

      // Map native errors to the shared JS contract from RNPingCore.
      val mappedError = when (error) {
        is MalformedURLException -> GenericError(
          type = ErrorType.ARGUMENT_ERROR,
          error = BrowserErrorCodes.BROWSER_OPEN_ERROR,
          message = error.message
        )
        else -> GenericError(
          type = ErrorType.INTERNAL_ERROR,
          error = BrowserErrorCodes.BROWSER_OPEN_ERROR,
          message = error?.message
        )
      }
      promise.reject(mappedError, error)
    }
  }

  /**
   * Reset any in-flight browser session (no-op on Android).
   */
  @JvmStatic
  fun reset() {
    // Android BrowserLauncher does not expose a public reset API.
  }

  /**
   * Resolve a native logger from the shared Core logger registry.
   *
   * @param id Logger handle identifier from JS.
   * @return Native logger instance, or null when missing/invalid.
   */
  private fun resolveLoggerFromCore(id: String?): Logger? {
    if (id.isNullOrBlank()) return null
    val handle = CoreRuntime.loggerRegistry.resolve(id) as? LoggerHandleContract ?: return null
    return handle.nativeLogger as? Logger
  }

  /**
   * Validate whether the target browser package is installed.
   */
  private fun isPackageInstalled(packageName: String): Boolean {
    return try {
      val pm = appContext?.packageManager ?: return false
      pm.getPackageInfo(packageName, 0)
      true
    } catch (e: Exception) {
      false
    }
  }

  /**
   * Parse and validate launch URL from an untrusted input string.
   *
   * Enforces:
   * - absolute URL with host
   * - `http` or `https` scheme only
   * - no embedded userinfo (credentials)
   */
  private fun parseLaunchUrl(rawUrl: String): URL {
    val uri = try {
      URI(rawUrl.trim())
    } catch (e: Exception) {
      throw MalformedURLException(e.message)
    }

    val scheme = uri.scheme?.lowercase()
    if (scheme != "http" && scheme != "https") {
      throw MalformedURLException("Unsupported URL scheme. Only HTTP and HTTPS URLs are supported.")
    }

    if (!uri.isAbsolute || uri.host.isNullOrBlank()) {
      throw MalformedURLException("Invalid URL. Provide an absolute HTTP(S) URL with a host.")
    }

    if (!uri.userInfo.isNullOrBlank()) {
      throw MalformedURLException("Unsupported URL format. User info is not allowed.")
    }

    return try {
      uri.toURL()
    } catch (e: Exception) {
      throw MalformedURLException(e.message)
    }
  }

  /**
   * Invoke Auth Tab builder APIs via reflection to avoid hard dependency on newer APIs.
   */
  private fun invokeAuthTabBuilderMethod(target: Any, name: String, vararg args: Any) {
    val method = target.javaClass.methods.firstOrNull {
      it.name == name && it.parameterTypes.size == args.size
    }
    if (method == null) {
      return
    }
    try {
      method.invoke(target, *args)
    } catch (e: Exception) {
    }
  }
}
